src/axom/primal/geometry/BezierPatch.hpp

```diff
@@ -0,0 +1,875 @@
+    }
+    else
+    {
+      m_weights.resize(sz_u, sz_v);
+
+      for(int t = 0; t < sz_u * sz_v; ++t)
+        m_weights(t / sz_v, t % sz_v) = weights[t];
+
+      SLIC_ASSERT(isValidRational());
+    }
+  }
+
+  /*!
+   * \brief Constructor for a Bezier Patch from an matrix of coordinates
+   *
+   * \param [in] pts a vector with (ord_u+1) * (ord_v+1) control points
+   * \param [in] ord_u The patch's polynomial order on the first axis
+   * \param [in] ord_v The patch's polynomial order on the second axis
+   * \pre order is greater than or equal to zero in each direction
+   *
+   */
+  BezierPatch(const CoordsMat& pts, int ord_u, int ord_v)
+  {
+    SLIC_ASSERT((ord_u >= 0) && (ord_v >= 0));
+
+    const int sz_u = utilities::max(0, ord_u + 1);
+    const int sz_v = utilities::max(0, ord_v + 1);
+
+    m_controlPoints.resize(sz_u, sz_v);
+    m_controlPoints = pts;
+
+    makeNonrational();
+  }
+
+  /*!
+   * \brief Constructor for a Rational Bezier Patch from a matrix
+   * of coordinates and weights
+   *
+   * \param [in] pts a vector with (ord_u+1) * (ord_v+1) control points
+   * \param [in] ord_u The patch's polynomial order on the first axis
+   * \param [in] ord_v The patch's polynomial order on the second axis
+   * \pre order is greater than or equal to zero in each direction
+   */
+  BezierPatch(const CoordsMat& pts,
+              const axom::Array<axom::Array<T>>& weights,
+              int ord_u,
+              int ord_v)
+  {
+    SLIC_ASSERT(ord_u >= 0 && ord_v >= 0);
+    SLIC_ASSERT(pts.shape()[0] == weights.shape()[0]);
+    SLIC_ASSERT(pts.shape()[1] == weights.shape()[1]);
+
+    const int sz_u = utilities::max(0, ord_u + 1);
+    const int sz_v = utilities::max(0, ord_v + 1);
+
+    m_controlPoints.resize(sz_u, sz_v);
+    m_controlPoints = pts;
+
+    m_weights.resize(sz_u, sz_v);
+    m_weights = weights;
+
+    SLIC_ASSERT(isValidRational());
+  }
+
+  /// Sets the order of the Bezier Patch
+  void setOrder(int ord_u, int ord_v)
+  {
+    m_controlPoints.resize(ord_u + 1, ord_v + 1);
+  }
+
+  /// Returns the order of the Bezier Patch on the first axis
+  int getOrder_u() const
+  {
+    return static_cast<int>(m_controlPoints.shape()[0]) - 1;
+  }
+
+  /// Returns the order of the Bezier Patch on the second axis
+  int getOrder_v() const
+  {
+    return static_cast<int>(m_controlPoints.shape()[1]) - 1;
+  }
+
+  /// Make trivially rational. If already rational, do nothing
+  void makeRational()
+  {
+    if(!isRational())
+    {
+      const int ord_u = getOrder_u();
+      const int ord_v = getOrder_v();
+
+      m_weights.resize(ord_u + 1, ord_v + 1);
+
+      for(int i = 0; i <= ord_u; ++i)
+        for(int j = 0; j <= ord_v; ++j) m_weights[i][j] = 1.0;
+    }
+  }
+
+  /// Make nonrational by shrinking array of weights
+  void makeNonrational() { m_weights.resize(0, 0); }
+
+  /// Use array size as flag for rationality
+  bool isRational() const { return (m_weights.size() != 0); }
```

Comment:
The following might be slightly clearer:
```suggestion
  bool isRational() const { return !m_weights.empty(); }
```